apps/src/codebridge/InfoPanel/ValidatedInstructions.tsx

```diff
@@ -0,0 +1,323 @@
+import {LevelStatus} from '@cdo/generated-scripts/sharedConstants';
+import commonI18n from '@cdo/locale';
+
+import {useCodebridgeContext} from '../codebridgeContext';
+import {appendSystemMessage} from '../redux/consoleRedux';
+
+import moduleStyles from '@codebridge/InfoPanel/styles/validated-instructions.module.scss';
+
+// By default we show the test and navigation buttons unless the URL parameter 'button-bar' is set.
+const SHOW_TEST_NAVIGATION_BUTTONS = !queryParams('button-bar');
+
+interface InstructionsProps {
+  /** Additional callback to fire before navigating to the next level. */
+  beforeNextLevel?: () => void;
+  /** If the instructions panel should be rendered vertically or horizontally. Defaults to vertical. */
+  layout?: 'vertical' | 'horizontal';
+  /**
+   * A callback when the user clicks on clickable text.
+   */
+  handleInstructionsTextClick?: (id: string) => void;
+  /** Optional classname for the container */
+  className?: string;
+}
+
+interface NavigationButtonProps {
+  showNavigation: boolean;
+  navigationText: string;
+  handleNavigation: () => void;
+  navigationIcon?: FontAwesomeV6IconProps;
+}
+
+/**
+ * Instructions panel with built-in validation feedback.
+ * This instructions panel shows the validation results for a level, if they exist,
+ * and provides navigation once validation has passed.
+ */
+const ValidatedInstructions: React.FunctionComponent<InstructionsProps> = ({
+  beforeNextLevel,
+  layout = 'vertical',
+  handleInstructionsTextClick,
+  className,
+}) => {
+  const {onRun, onStop} = useCodebridgeContext();
+  const dialogControl = useDialogControl();
+
+  const instructionsText = useAppSelector(
+    state => state.lab.levelProperties?.longInstructions
+  );
+  const hasNextLevel = useSelector(state => nextLevelId(state) !== undefined);
+  const {hasConditions, satisfied} = useAppSelector(
+    state => state.lab.validationState
+  );
+  const predictSettings = useAppSelector(
+    state => state.lab.levelProperties?.predictSettings
+  );
+  const predictResponse = useAppSelector(state => state.predictLevel.response);
+  const predictAnswerLocked = useAppSelector(isPredictAnswerLocked);
+  const hasRun = useAppSelector(state => state.lab2System.hasRun);
+  const hasSubmitted = useAppSelector(
+    state => getCurrentLevel(state)?.status === LevelStatus.submitted
+  );
+  const isSubmittable = useAppSelector(
+    state => state.lab.levelProperties?.submittable
+  );
+  const source = useAppSelector(
+    state => state.lab2Project.projectSource?.source
+  ) as MultiFileSource | undefined;
+
+  const appType = useAppSelector(state => state.lab.levelProperties?.appName);
+  const isTesting = useAppSelector(state => state.lab2System.isTesting);
+  const isLoadingEnvironment = useAppSelector(
+    state => state.lab2System.loadingCodeEnvironment
+  );
+
+  const dispatch = useAppDispatch();
+
+  const {theme} = useContext(ThemeContext) || 'dark';
```

Comment:
I don't think we need the `|| 'dark'`. It looks like the `ThemeContext` provides a default theme, which is `dark`.

Reply:
I'll fix this in the follow-up! Don't want to wait for drone if I don't have to.